internal/service/componentdescriptor/service_test.go

```diff
@@ -0,0 +1,303 @@
+		cdBytes, err := compdesc.Encode(cd)
+		require.NoError(t, err)
+		tarBytes := wrapAsTar(t, "invalid-name", cdBytes)
+		mockLayer := static.NewLayer(tarBytes, componentDescriptorMediaType)
+
+		repo := mockOCIRepository{
+			getConfigResult: []byte(testComponentDescriptorConfig),
+			pullLayerResult: mockLayer,
+		}
+
+		svc, err := componentdescriptor.NewService(&repo)
+		require.NoError(t, err)
+
+		ocmi, err := ocmidentity.New(testComponentName, testComponentVersion)
+		require.NoError(t, err)
+
+		// when
+		result, err := svc.GetComponentDescriptor(t.Context(), *ocmi)
+
+		// then
+		require.Error(t, err)
+		assert.Nil(t, result)
+		require.ErrorIs(t, err, componentdescriptor.ErrNotFoundInTar)
+		assert.Contains(t, err.Error(), "not found in TAR archive")
+		assert.Contains(t, err.Error(), "failed to extract data of file=\""+
+			componentdescriptor.ComponentDescriptorFileName+"\"")
+	})
+
+	t.Run("should fail on component descriptor decoding error", func(t *testing.T) {
+		// given
+		cd := compdesc.New(testComponentName, testComponentVersion)
+		compdesc.RegisterScheme(&compdescv2.DescriptorVersion{})
+		cdBytes, err := compdesc.Encode(cd)
+		require.NoError(t, err)
+		// introduce an error
+		cdBytes = bytes.ReplaceAll(cdBytes, []byte("[]"), []byte("!}"))
+		tarBytes := wrapAsTar(t, componentdescriptor.ComponentDescriptorFileName, cdBytes)
+		mockLayer := static.NewLayer(tarBytes, componentDescriptorMediaType)
+
+		repo := mockOCIRepository{
+			getConfigResult: []byte(testComponentDescriptorConfig),
+			pullLayerResult: mockLayer,
+		}
+
+		svc, err := componentdescriptor.NewService(&repo)
+		require.NoError(t, err)
+
+		ocmi, err := ocmidentity.New(testComponentName, testComponentVersion)
+		require.NoError(t, err)
+
+		// when
+		_, err = svc.GetComponentDescriptor(t.Context(), *ocmi)
+
+		// then
+		require.Error(t, err)
+		assert.Contains(t, err.Error(), "failed to decode component descriptor fetched from ocm artifact with name=\""+
+			testComponentName+"\" and version=\""+testComponentVersion+"\"")
+	})
+}
+
+type mockOCIRepository struct {
+	getConfigName   string
+	getConfigTag    string
+	pullLayerName   string
+	pullLayerTag    string
+	pullLayerDigest string
+	getConfigResult []byte
+	getConfigError  error
+	pullLayerResult containerregistryv1.Layer
+	pullLayerError  error
+}
+
+func (m *mockOCIRepository) GetConfigFile(ctx context.Context, name, tag string) ([]byte, error) {
+	m.getConfigName = name
+	m.getConfigTag = tag
+	if m.getConfigError != nil {
+		return nil, m.getConfigError
+	}
+	return m.getConfigResult, nil
+}
+
+func (m *mockOCIRepository) PullLayer(ctx context.Context, name, tag, digest string) (
+	containerregistryv1.Layer, error,
+) {
+	m.pullLayerName = name
+	m.pullLayerTag = tag
+	m.pullLayerDigest = digest
+	if m.pullLayerError != nil {
+		return nil, m.pullLayerError
+	}
+	return m.pullLayerResult, nil
+}
+
+func wrapAsTar(t *testing.T, fileName string, data []byte) []byte {
```

Comment:
This is the bad concesquence that you deeply coupled tar related operation into `GetComponentDescriptor`, a good unit test doesn't require such low level detail. The reason I mentioned here https://github.com/kyma-project/lifecycle-manager/pull/2736#discussion_r2435890501